src/QAT/qat/qat.py

```diff
@@ -74,8 +75,13 @@ def execute_with_metrics(
 ):
     """ Execute file path or code blob. """
 
-    frontend: LanguageFrontend = fetch_frontend(path_or_str)
-    return _execute_with_metrics(frontend, path_or_str, hardware, compiler_config)
+    total_execution_timer = log_duration("Program execution completed. Total runtime took {} seconds.")
+    with total_execution_timer:
+        frontend: LanguageFrontend = fetch_frontend(path_or_str)
+        results, metrics = _execute_with_metrics(frontend, path_or_str, hardware, compiler_config)
+    metrics.record_metric(MetricsType.TotalDuration, total_execution_timer.duration)
+
+    return results, metrics.as_dict()
```

Comment:
This should all be moved to `_execute_with_metrics`. As a side note, you changed the return value of every other entry-point API - `execute_qir_x` etc.